rules/firestore.json

```diff
@@ -35,50 +36,188 @@ service cloud.firestore {
+            return verifiedKeptSameValue || verifiedSetToTrueAfterVerification
+        }
+        match /parts/{id} {
+            // Anybody can create a part
+            allow create: if userIsVerified();
+            // The owner of a part can delete it, if it has not started
+            allow delete: if userIsVerified() &&
+                             creatorOf(id).id == request.auth.uid &&
+                             partHasNotStarted(resource.data);
+            // Anyone can delete a part if the owner has timed out and the part has not started
+            allow delete: if userIsVerified() &&
+                             hasTimedOut(creatorOf(id)) &&
+                             partHasNotStarted(resource.data);
+            // A number of fields are readonly:
+            // - typeGame
+            // - playerZero/playerOne once the part has started
+            // - beginning once the part has started
+            // the part is considered to be started as soon as beginning is set
+            // For any other fields, both players can modify them, with some restrictions:
+            // - lastUpdate must be updated monotonously by the corresponding player
+            // - turn can only be made +1 (usual turn), -1 (take back), or -2 (take back)
+            allow update: if userIsVerified() &&
+                             preservesField('typeGame') &&
+                             (preservesField('playerZero') || partHasNotStarted(resource.data)) &&
+                             (preservesField('playerOne') || partHasNotStarted(resource.data)) &&
+                             (preservesField('beginning') || partHasNotStarted(resource.data)) &&
+                             (preservesField('turn') || safeTurnUpdate(resource.data.turn, request.resource.data.turn)) &&
+                             (safeLastUpdate(resource.data.lastUpdate, request.resource.data.lastUpdate)) &&
+                             (userIsPlayer(resource.data) || (partHasNotStarted(resource.data) && userIsChosenOpponentInJoiner(id)));
+
+            // All verified user can read parts (needed for the lobby)
+            allow read: if userIsVerified();
+        }
+        function userIsChosenOpponentInJoiner(partId) {
+            let joiner = get(/databases/$(database)/documents/joiners/$(partId)).data;
+            return hasField(joiner, 'chosenOpponent') && joiner.chosenOpponent.id == request.auth.uid;
+        }
+        function userIsPlayer(part) {
+            let userIsPlayerZero = part.playerZero.id == request.auth.uid;
+            let userIsPlayerOne = hasField(part, 'playerOne') && part.playerOne.id == request.auth.uid;
+            return userIsPlayerZero || userIsPlayerOne;
+        }
+        function safeTurnUpdate(oldTurn, newTurn) {
+            return (newTurn == oldTurn + 1) || (newTurn == oldTurn - 1) || (newTurn == oldTurn - 2);
```

Comment:
idéalement, avec le commentaire précédent ou on aurait
allow update if : isMove();
allow update if : isTakeBackAccepted(); ça serait plus complet, là y a de la place pour du bloubiboulga ou j'append dans listMoves et je décrémente dans turn (tant qu'on est à penser à wtfuckisme des canailles d'intraweb!)

Reply:
normalement c'est déjà plus clair, mais pas moyen de voir si listMoves est bien updaté (pour rappel, == est tout pourri ici, donc listMoves étant une liste c'est très casse couille), donc pas moyen de distinguer isMove de isSomethingElse, donc on fait au mieux avec ce qu'on a :shrug:

Reply:
isMove j'entendrais par là que:
1. le tour a été incrémenté
2. la longeur de la liste de mouvement a été incrémentée
3. éventuellement (sauf au premier mouvement) le temps du joueur adverse a été diminué
4. évidemment comme tout update: lastUpdate est changé adéquatos
5. un éventuel changement du score
6. et les checks d'authorisation que tu joue bien à TON tour dans une partie non finie
7. ah et un éventuel changement de status de la partie ('eheh, j'ai gagné là, coquin!')

Reply:
Alors j'ai un peu refactoré tout ça pour splitter move/non move.
Les points non gérés pour le moment:
2 est impossible (on peut savoir que listMoves est une liste, mais on peut pas accéder à sa taille !)
5 n'est pas une restriction !
7 bah comment tu fais pour changer le statut de la partie quand quelqu'un gagne si c'est interdit ? :D

Reply:
5, certes mais rapellons nous que seul un move ou un tackBack peut modifier le score, right !
7, qui a parlé d'interdire cela :p ? On sais que c'est une des attaques rigolotes les plus probable et maline :P
ça aurais été drôle de rester en backless et dire que c'est à l'adversaire de confirmer "eh ! c'était illégal ça! ALERTE ! Je joue contre un malicieux malicien !!!"

Reply:
Le problème de 7 c'est que faut que ça soit solide des deux côtés (je pense qu'on a déjà eu cette discussion): inversément à ton scénario, tu peux en tant qu'user normal me dire « j'ai gagné » et moi je peux refuser en tant que malicieux ! La seule façon de vérifier cela en ne trustant pas les users et d'avoir une source de vérité (= les rules du backend)

Reply:
Et j'ai ajouté 5 au ticket backend

Reply:
Attends je comprends pas, vérifier qu'une victoire ne peut venir qu'avec un move, pas juste isolément, ça nécessite pas le backend, Et qu'un score peut être modifié qu'avec un take-back et un move pareil, no need le back ?

Reply:
mais comment tu t'assures que j'envoie pas une victoire quand j'ai pas gagné ?

Reply:
ça c'est une autre question, mais dans le test des victoires, le test que ça va avec un mouvement est quand même obligatoire, mais clairement le check non faisable susmentionné est plus utile xD mais cuilà est ticketté déjà !

Reply:
J'allais ajouter preservesField(winner) et loser, mais c'n'est pas vrai ! C'est soit en cas de mouvement gagnant, soit en cas de resign. C'est loin d'être trivial tout ça :D Mais ajouté et testé

Reply:
ah mince oui, c'est vrai, t'as testé que les nonMove touche pas, mais en effet les resign existent, et bah du coup, ils sont cassés là ?

Reply:
not anymore